### PR TITLE
Debian and Ubuntu image support

### PR DESCRIPTION
**Pull request by leki75**

### CODE REVIEW

**Comment by leki75** [comment]
Do not merge yet! Vagrant and OpenStack validation in progress.

File: Vagrantfile [hunk 1]
```diff
@@ -34,63 +34,67 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
...
-      #   :ambari_baseurl => "http://s3.amazonaws.com/dev.hortonworks.com/ambari/ubuntu14/2.x/BUILDS/2.4.1.0-22/",
-      #   :ambari_key => "B9733A7A07513CAD",
-      #   :salt_repo => "http://repo.saltstack.com/apt/ubuntu/14.04/amd64/latest"
-      # },
+      {
+        :hostname => "centos6-vagrant",
+        :box => "centos/6",
+        :ram => 1536,
+        :cpu => 2,
+        :salt_repo => "salt-repo-2016.11-6.el.repo",
```
Comment (tibkiss) [inline Vagrantfile]: Shouldn't this bumped to 2017?

File: Vagrantfile [hunk 2]
```diff
@@ -34,63 +34,67 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
...
+        :cpu => 2,
+        :salt_repo => "salt-repo-2016.11-5.debian7.list",
+        :custom_image_type => "hortonworks"
+      },
+      {
+        :hostname => "precise-vagrant",
+        :box => "ubuntu/precise64",
+        :ram => 1536,
+        :cpu => 2,
+        :salt_repo => "salt-repo-2016.11-3.ubuntu12.list",
```
Comment (tibkiss) [inline Vagrantfile]: as above

Comment (leki75) [reply Vagrantfile]: No newer repository is available for Ubuntu 12.04.
http://repo.saltstack.com/apt/ubuntu/12.04/amd64/archive/

File: Vagrantfile (hunk 1 above)
Comment (leki75) [reply Vagrantfile]: We faced some problems with newer salt versions on centos6. It seemed to be the most stable one.

File: Vagrantfile (hunk 2 above)
Comment (lacikaaa) [reply Vagrantfile]: Ubuntu 12.04 is not supported by Canonical: https://wiki.ubuntu.com/Releases

Comment (leki75) [reply Vagrantfile]: You are right but HDP is still supported on Ubuntu 12.04. I have the same problem with Debian 7 (support ends on 31st May 2018).

**Comment by leki75** [comment]
Build succeeded on vagrant and on openstack for centos6, centos7, debian7 and ubuntu16.
Centos6 and centos7 were successfully deployed with cloudbreak 1.16.

**Comment by tibkiss** [comment]
Re: removing ubuntu12. I'd keep that OS since it is a valid sustaining use-case (need to support outdated os-es and hdp, ambari versions).

File: Vagrantfile (hunk 1 above)
Comment (tibkiss) [reply Vagrantfile]: makes sense, thanks

File: saltstack/base/salt/cloud-init/init.sls [hunk 3]
```diff
@@ -1,38 +1,17 @@
...
-    - unless: ls /etc/waagent.conf
-
 
 preserve_hostname_false:
   file.replace:
     - name: /etc/cloud/cloud.cfg
     - pattern: "^preserve_hostname.*"
     - repl: "preserve_hostname: true"
     - append_if_not_found: True
-    - unless: ls /etc/waagent.conf
```
Comment (tibkiss) [inline saltstack/base/salt/cloud-init/init.sls]: What's the rationale removing these?

File: saltstack/base/salt/prerequisites/disable_ipv6.sls [hunk 4]
```diff
@@ -10,9 +10,12 @@ net.ipv6.conf.lo.disable_ipv6:
   sysctl.present:
     - value: 1
 
-net.ipv6.conf.eth0.disable_ipv6:
-  sysctl.present:
-    - value: 1
+# On newer OSes (like Ubuntu Xenial) the naming of devices
```
Comment (tibkiss) [inline saltstack/base/salt/prerequisites/disable_ipv6.sls]: Please add a TODO or FIXME here

File: saltstack/base/salt/prerequisites/init.sls [hunk 5]
```diff
@@ -13,6 +13,8 @@ include:
   - {{ slspath }}.cert-tool
   - {{ slspath }}.dnsmasq
   - {{ slspath }}.user
+  - {{ slspath }}.selinux
```
Comment (tibkiss) [inline saltstack/base/salt/prerequisites/init.sls]: This will not work on docker build. Shouldn't we put this into else part of the docker check at line 6?

File: saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh [hunk 6]
```diff
@@ -27,6 +27,16 @@ wait_for_authorized_keys() {
 }
 
 setup_tmp_ssh() {
+
+  if ! id ${SSH_USER}; then
```
Comment (tibkiss) [inline saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh]: I'm not sure why is this needed here as we have similar functionality in the prerequisites:
https://github.com/hortonworks/cloudbreak-images/blob/master/saltstack/base/salt/prerequisites/user.sls

File: saltstack/final/salt/validate/init.sls [hunk 7]
```diff
@@ -2,17 +2,13 @@ read_java_home:
...
     - name: /etc/profile.d/java.sh
     - failhard: True
-  cmd.run:
-    - name: 'source /etc/profile.d/java.sh'
-
 
 check_java_javahome:
   cmd.run:
-    - name: 'source /etc/profile.d/java.sh; $JAVA_HOME/bin/java -version'
+    - name: 'source /etc/profile.d/java.sh && $JAVA_HOME/bin/java -version'
```
Comment (tibkiss) [inline saltstack/final/salt/validate/init.sls]: This is a very nice catch!

File: saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh (hunk 6 above)
Comment (leki75) [reply saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh]: This patch is based on rc-1.16 where those modifications do not exist. Removing it.

File: saltstack/base/salt/cloud-init/init.sls (hunk 3 above)
Comment (leki75) [reply saltstack/base/salt/cloud-init/init.sls]: I have no information about this. Heritage from @kozmagabor

File: saltstack/base/salt/prerequisites/disable_ipv6.sls (hunk 4 above)
Comment (leki75) [reply saltstack/base/salt/prerequisites/disable_ipv6.sls]: Sure

**Comment by tibkiss** [comment]
@kozmagabor : could you please chime in regarding the waagent related changes?

**Comment by tibkiss** [comment]
Talked with @kozmagabor about waagent change. He has moved out the waagent check to the top level - which makes.

**Comment by tibkiss** [comment]
LGTM

File: saltstack/base/salt/cloud-init/init.sls (hunk 3 above)
Comment (lacikaaa) [reply saltstack/base/salt/cloud-init/init.sls]: as I can see it is moved [here](https://github.com/hortonworks/cloudbreak-images/pull/178/files#diff-2ca6b90a7d998c209b535590afd5ac13R21)

File: saltstack/base/salt/performance/thp.sls [hunk 8]
```diff
@@ -1,14 +1,17 @@
+{% if grains['init'] in [ 'upstart', 'sysvinit'] %}
 /etc/init.d/disable-thp:
   file.managed:
-    - source: salt://{{ slspath }}/etc/init.d/disable-thp
+    - source:
+      - salt://{{ slspath }}/etc/init.d/disable-thp.{{ grains['os'] | lower }}
```
Comment (lacikaaa) [inline saltstack/base/salt/performance/thp.sls]: This only exists for debian and also seems to me that it does the same as [disable-thp](https://github.com/hortonworks/cloudbreak-images/blob/master/saltstack/base/salt/performance/etc/init.d/disable-thp) already does

File: saltstack/base/salt/unbound/init.sls [hunk 9]
```diff
@@ -52,16 +54,41 @@ config_unbound_server:
...
+    - name: /etc/unbound/local.d/block-example.com.conf
+    - source: salt://{{ slspath }}/etc/unbound/local.d/block-example.com.conf
+    - mode: 644
+
+config_unbound_example_conf_d:
+  file.managed:
+    - user: root
+    - group: root
+    - name: /etc/unbound/conf.d/example.com.conf
+    - source: salt://{{ slspath }}/etc/unbound/conf.d/example.com.conf
```
Comment (lacikaaa) [inline saltstack/base/salt/unbound/init.sls]: Example configs should not be activated by default. If you modify the file extension to anything else than `.conf` then it wouldn't be applied by unbound.

File: saltstack/optional/salt/oracle-java/init.sls [hunk 10]
```diff
@@ -13,46 +13,33 @@ remove_openjdk:
...
     - mode: 755
+    - contents: |
+        export JAVA_HOME={{ pillar['JAVA_HOME'] }}
 
-{% if grains['init'] == 'systemd' %}
 set_java_home_systemd:
   file.replace:
     - name: /etc/systemd/system.conf
     - pattern: \#*DefaultEnvironment=.*
     - repl: DefaultEnvironment=JAVA_HOME=/usr/java/default
```
Comment (lacikaaa) [inline saltstack/optional/salt/oracle-java/init.sls]: This is not consistent with `set_java_home_user:` where you use `{{ pillar['JAVA_HOME'] }}` as the source for JAVA_HOME.

File: saltstack/optional/salt/oracle-java/init.sls [hunk 11]
```diff
@@ -13,46 +13,33 @@ remove_openjdk:
...
-{% if grains['os_family'] == 'RedHat' %}
 download_oracle_jdk:
   {% set jdk_url = salt['environ.get']('ORACLE_JDK8_URL_RPM') %}
-  {% set download_jdk_cmd = ['\'wget --no-cookies --no-check-certificate --header "Cookie: gpw_e24=http%3A%2F%2Fwww.oracle.com%2F; oraclelicense=accept-securebackup-cookie" "',
+  {% set download_jdk_cmd = ['\'wget -c -q --no-cookies --no-check-certificate --header "Cookie: gpw_e24=http%3A%2F%2Fwww.oracle.com%2F; oraclelicense=accept-securebackup-cookie" "',
                              jdk_url,
                              '" -O /tmp/oracle_jdk_install.rpm\'' ] %}
   cmd.run:
     - name: {{ download_jdk_cmd | join('') }}
-    - unless: test -e /tmp/oracle_jdk_install.rpm
```
Comment (lacikaaa) [inline saltstack/optional/salt/oracle-java/init.sls]: With this option it was possible to place an rpm there and install it without downloading. Is it really necessary to remove?

File: saltstack/optional/pillar/top.sls [hunk 12]
```diff
@@ -1,3 +1,4 @@
 optional:
   '*':
-    - dummy
\ No newline at end of file
+    - dummy
```
Comment (lacikaaa) [inline saltstack/optional/pillar/top.sls]: If you add oracle-java then dummy should be removed.

File: saltstack/base/salt/performance/thp.sls (hunk 8 above)
Comment (leki75) [reply saltstack/base/salt/performance/thp.sls]: It is also installed for centos6 as it uses sysv init system. Debian7 has some kind of dependency management that is why BEGIN INIT INFO and END INIT INFO comments are there. As the 2 files are similar it is possible to merge them together.

File: saltstack/optional/salt/oracle-java/init.sls (hunk 11 above)
Comment (leki75) [reply saltstack/optional/salt/oracle-java/init.sls]: With the -c option of wget we get the same functionality.

File: saltstack/optional/pillar/top.sls (hunk 12 above)
Comment (leki75) [reply saltstack/optional/pillar/top.sls]: Corrected, thanks.

File: saltstack/optional/salt/oracle-java/init.sls (hunk 10 above)
Comment (leki75) [reply saltstack/optional/salt/oracle-java/init.sls]: You are right. Corrected.

File: saltstack/config/minion [hunk 13]
```diff
@@ -1,6 +1,7 @@
 file_client: local
 requests_lib: True
 backend_requests: True
+pillarenv_from_saltenv: True
```
[the remark this answers is not in the document]
Comment (leki75) [reply saltstack/config/minion]: In scripts/salt-setup.sh saltenv parameter is used.

File: saltstack/base/salt/unbound/init.sls (hunk 9 above)
Comment (leki75) [reply saltstack/base/salt/unbound/init.sls]: Fixed

File: packer.json [hunk 14]
```diff
@@ -34,13 +34,16 @@
     "ssh_password": "S3cr3t",
     "atlas_artifact": "cloudbreak",
     "atlas_artifact_type": "{{ env `ATLAS_ARTIFACT_TYPE` }}",
+    "atlas_meta_os_distribution_id": "{{ env `ATLAS_META_OS_DISTRIBUTION_ID` }}",
```
[the remark this answers is not in the document]
Comment (leki75) [reply packer.json]: Fixed

File: saltstack/base/salt/unbound/init.sls (hunk 9 above)
Comment (leki75) [reply saltstack/base/salt/unbound/init.sls]: Done
